File: geometry.js
import {XMLParser, XMLValidator} from 'fast-xml-parser';
export const SURFACES=['unknown','asphalt','gravel','trail'];
export const num=value=>{if(value===null||value===undefined||value==='')throw Error('Ange ett giltigt tal.');const n=Number(value);if(!Number.isFinite(n))throw Error('Ange ett giltigt tal.');return n;};
export function distance(a,b){const rad=Math.PI/180,h=Math.sin((b[0]-a[0])*rad/2)**2+Math.cos(a[0]*rad)*Math.cos(b[0]*rad)*Math.sin((b[1]-a[1])*rad/2)**2;return 6371008.8*2*Math.asin(Math.min(1,Math.sqrt(h)));}
const arr=x=>x===undefined?[]:Array.isArray(x)?x:[x];
export function parseGPX(raw){
 const xml=raw.toString('utf8');if(/<!DOCTYPE|<!ENTITY/i.test(xml)||XMLValidator.validate(xml)!==true)throw Error('Filen kunde inte läsas som GPX. Exportera den på nytt.');
 const tree=new XMLParser({ignoreAttributes:false,removeNSPrefix:true,parseTagValue:false,processEntities:true}).parse(xml);
 if(!tree.gpx)throw Error('Det här är inte en GPX-fil.');
 const tracks=arr(tree.gpx.trk),routes=arr(tree.gpx.rte);let groups=tracks.flatMap(t=>arr(t.trkseg).map(s=>arr(s.trkpt)));if(!groups.length)groups=routes.map(r=>arr(r.rtept));
 const points=[];let total=0,gain=0,loss=0,known=0,edges=0;
 groups.forEach((group,segment)=>{let prev=null,anchor=null;for(const el of group){const lat=num(el['@_lat']),lon=num(el['@_lon']);if(Math.abs(lat)>90||Math.abs(lon)>180)throw Error('Ogiltiga koordinater.');const ele=el.ele===undefined?null:num(el.ele);const p=[lat,lon,ele,0,segment];if(prev){total+=distance(prev,p);edges++;if(prev[2]!==null&&ele!==null)known++;}p[3]=Math.round(total*1000)/1000;points.push(p);if(ele===null)anchor=null;else if(anchor===null)anchor=ele;else if(Math.abs(ele-anchor)>=3){gain+=Math.max(0,ele-anchor);loss+=Math.max(0,anchor-ele);anchor=ele;}prev=p;if(points.length>60000)throw Error('Filen har fler än 60 000 punkter. Förenkla GPX-spåret först.');}});
 if(points.length<2||total<1)throw Error('Filen behöver minst två olika spårpunkter.');
 const heights=points.filter(p=>p[2]!==null).map(p=>p[2]);
 return {points,distance:Math.round(total*1000)/1000,ascent:known?Math.round(gain):null,descent:known?Math.round(loss):null,elevation_coverage:Math.round(known/Math.max(1,edges)*1000)/1000,min_elevation:heights.length?Math.min(...heights):null,max_elevation:heights.length?Math.max(...heights):null,gpx_name:String(tracks[0]?.name??routes[0]?.name??'').slice(0,120),segments:new Set(points.map(p=>p[4])).size};
}
export function paint(spans,start,end,kind,total,source='manual'){
 start=num(start);end=num(end);if(!SURFACES.includes(kind)||start<0||start>=end||end>total+.01)throw Error('Välj ett intervall inom rutten och ett giltigt underlag.');end=Math.min(end,total);const result=[];
 for(const s of spans){if(s.end<=start||s.start>=end)result.push({...s});else{if(s.start<start)result.push({...s,end:start});result.push({start:Math.max(start,s.start),end:Math.min(end,s.end),kind,source});if(s.end>end)result.push({...s,start:end});}}
 const merged=[];for(const s of result){const last=merged.at(-1);if(last&&last.kind===s.kind&&last.source===s.source)last.end=s.end;else merged.push(s);}return merged;
}
export function sample(points,limit=220){const step=Math.max(1,Math.ceil(points.length/limit)),chosen=new Set([0,points.length-1]);for(let i=0;i<points.length;i+=step)chosen.add(i);for(let i=1;i<points.length;i++)if(points[i][4]!==points[i-1][4]){chosen.add(i-1);chosen.add(i);}return [...chosen].sort((a,b)=>a-b).map(i=>points[i]);}
const escapeXML=v=>String(v??'').replace(/[&<>"']/g,c=>({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&apos;'}[c]));
export function exportGPX(r){let xml='<?xml version="1.0" encoding="UTF-8"?><gpx xmlns="http://www.topografix.com/GPX/1/1" version="1.1" creator="Alvins Ruttbank">';for(const c of r.cafes)xml+=`<wpt lat="${c.lat}" lon="${c.lon}"><name>${escapeXML(c.name)}</name><desc>${escapeXML(c.note)}</desc></wpt>`;xml+=`<trk><name>${escapeXML(r.name)}</name><desc>${escapeXML(r.description)}</desc>`;let old=null;for(const [lat,lon,ele,,seg] of r.points){if(seg!==old){if(old!==null)xml+='</trkseg>';xml+='<trkseg>';old=seg;}xml+=`<trkpt lat="${lat}" lon="${lon}">${ele!==null?`<ele>${ele}</ele>`:''}</trkpt>`;}return xml+'</trkseg></trk></gpx>';}

File: geometry.py
"""GPX parsing, distance, elevation and surface interval operations."""
import math
from bisect import bisect_right
from defusedxml import ElementTree as SafeET
from xml.etree import ElementTree as ET

SURFACES = ('unknown', 'asphalt', 'gravel', 'trail', 'other')

def number(value):
    try:
        n = float(value)
    except (ValueError, TypeError):
        raise ValueError('Ange ett giltigt tal.')
    if not math.isfinite(n):
        raise ValueError('Ogiltigt tal i filen.')
    return n

def distance(a, b):
    lat1, lat2 = math.radians(a[0]), math.radians(b[0])
    dlat, dlon = lat2-lat1, math.radians(b[1]-a[1])
    h = math.sin(dlat/2)**2+math.cos(lat1)*math.cos(lat2)*math.sin(dlon/2)**2
    return 6371008.8*2*math.asin(min(1, math.sqrt(h)))

def parse_gpx(raw):
    try:
        root = SafeET.fromstring(raw)
    except Exception:
        raise ValueError('Filen kunde inte läsas som GPX. Exportera den på nytt och försök igen.')
    if root.tag.split('}')[-1] != 'gpx':
        raise ValueError('Det här är inte en GPX-fil.')
    groups = root.findall('.//{*}trkseg')
    if not groups:
        groups = root.findall('./{*}rte')
    points, total, gain, loss = [], 0., 0., 0.
    known_edges = 0
    for segment, group in enumerate(groups):
        prev, anchor = None, None
        nodes = [p for p in group if p.tag.split('}')[-1] in ('trkpt','rtept')]
        for el in nodes:
            try:
                lat, lon = number(el.attrib['lat']), number(el.attrib['lon'])
                if not -90 <= lat <= 90 or not -180 <= lon <= 180:
                    raise ValueError()
                elev = el.find('{*}ele')
                height = number(elev.text) if elev is not None and elev.text else None
            except (KeyError, ValueError, TypeError):
                raise ValueError('GPX-filen innehåller ogiltiga koordinater eller höjdvärden.')
            p = [lat, lon, height, 0., segment]
            if prev:
                total += distance(prev, p)
                if prev[2] is not None and height is not None:
                    known_edges += 1
            p[3] = round(total, 3)
            points.append(p)
            if height is None:
                anchor = None
            elif anchor is None:
                anchor = height
            elif abs(height-anchor) >= 3:
                gain += max(0, height-anchor)
                loss += max(0, anchor-height)
                anchor = height
            prev = p
            if len(points) > 60000:
                raise ValueError('Filen har fler än 60 000 punkter. Förenkla GPX-spåret först.')
    if len(points) < 2 or total < 1:
        raise ValueError('Filen behöver ett spår eller en rutt med minst två olika punkter.')
    label = root.find('.//{*}trk/{*}name')
    if label is None:
        label = root.find('.//{*}rte/{*}name')
    heights = [p[2] for p in points if p[2] is not None]
    edge_count = sum(a[4] == b[4] for a,b in zip(points, points[1:]))
    return {'points': points, 'distance': round(total, 3),
            'ascent': round(gain) if known_edges else None,
            'descent': round(loss) if known_edges else None,
            'elevation_coverage': round(known_edges/max(1, edge_count), 3),
            'min_elevation': min(heights) if heights else None,
            'max_elevation': max(heights) if heights else None,
            'gpx_name': (label.text or '').strip()[:120] if label is not None else '',
            'segments': len(set(p[4] for p in points))}

def paint(spans, start, end, kind, total, source='manual'):
    start, end = number(start), number(end)
    if kind not in SURFACES or not 0 <= start < end <= total+.01:
        raise ValueError('Välj ett giltigt underlag och ett intervall inom rutten.')
    end = min(end, total)
    result = []
    for span in spans:
        a,b = span['start'],span['end']
        if b <= start or a >= end:
            result.append(span.copy())
        else:
            if a < start: result.append({**span,'end':start})
            result.append({'start':max(a,start),'end':min(b,end),'kind':kind,'source':source})
            if b > end: result.append({**span,'start':end})
    merged=[]
    for span in result:
        if merged and merged[-1]['kind']==span['kind'] and merged[-1]['source']==span['source']:
            merged[-1]['end']=span['end']
        else: merged.append(span)
    return merged

def sample(points, limit=220):
    step=max(1,math.ceil(len(points)/limit))
    chosen={0,len(points)-1,*range(0,len(points),step)}
    for i in range(1,len(points)):
        if points[i][4]!=points[i-1][4]:chosen.update((i-1,i))
    return [points[i] for i in sorted(chosen)]

def export_gpx(route):
    ns='http://www.topografix.com/GPX/1/1'
    ET.register_namespace('',ns)
    root=ET.Element('{'+ns+'}gpx',version='1.1',creator='Alvins Ruttbank')
    track=ET.SubElement(root,'{'+ns+'}trk')
    ET.SubElement(track,'{'+ns+'}name').text=route['name']
    ET.SubElement(track,'{'+ns+'}desc').text=route['description']
    old=None
    for lat,lon,ele,_,seg in route['points']:
        if old!=seg:
            group=ET.SubElement(track,'{'+ns+'}trkseg');old=seg
        p=ET.SubElement(group,'{'+ns+'}trkpt',lat=str(lat),lon=str(lon))
        if ele is not None: ET.SubElement(p,'{'+ns+'}ele').text=str(ele)
    # GPX schema requires waypoints before tracks.
    for cafe in reversed(route.get('cafes',[])):
        p=ET.Element('{'+ns+'}wpt',lat=str(cafe['lat']),lon=str(cafe['lon']))
        ET.SubElement(p,'{'+ns+'}name').text=cafe['name']
        ET.SubElement(p,'{'+ns+'}desc').text=cafe.get('note','')
        root.insert(0,p)
    return ET.tostring(root,encoding='utf-8',xml_declaration=True)
